doc/build/core/temporal.rst

```diff
@@ -0,0 +1,238 @@
+only a brief introduction to these new temporal features; for a more complete
+explanation, refer to `TF`_.
+
+.. _TF: https://cs.ulb.ac.be/public/_media/teaching/infoh415/tempfeaturessql2011.pdf
+
+These temporal features are split into system versioning and application
+versioning. System versioning describes row history logging that is handled
+automatically by the backend, useful for producing auditable tables and storing
+history without triggers. Historical rows are (usually) held in the same table,
+and are hidden unless explicitly requested using ``FOR SYSTEM TIME`` in
+``SELECT`` statements.
+
+Application versioning describes roww with a time period indicating its valid
+time, manually managed by the application rather than the by the backend. This
+is useful for situations where something might be valid for only a specific time
+period, e.g. insurance policies or a period of employment.
+
+System versioning is currently supported only by MariaDB and Microsoft SQL
+Server. Application versioning is only supported by MariaDB.
+
+At the moment, SQLAlchemy supports all temporal DML (``CREATE TABLE``)
+constructs for all database backends that implement it; however, it does not
+support DML constucts. These include  ``SELECT [...] FOR SYSTEM_TIME [...]`` for
+system versioning and ``SELECT [...] WHERE my_period CONTAINS [...]`` with
+application versioning - these actions can mostly be accomplished using
+:meth:`_expression.Select.with_hint` for system versioning, and and text
+:meth:`_orm.Query.where`/ :meth:`_orm.Query.filter` clauses for application
+versioning.
+
+
+Working with Application-Time Periods
+-------------------------------------
+Application-time period schemas can be represented in SQLAlchemy using the
+:class:`Period` schema item, or its alias :class:`ApplicationTimePeriod`. All
+that is required in SQL DDL to add application versioning to a table is to
+include a named ``PERIOD FOR`` construct. This is described similarly in
+SQLAlchemy, using something like the following:
+
+.. code-block:: python
+
+    t = Table(
+        "t",
+        metadata,
+        Column("id", Integer),
+        Column("start_ts", TIMESTAMP),
+        Column("end_ts", TIMESTAMP),
+        Period("my_period", "start_ts", "end_ts"),
+    )
+
+The above will compile to:
+
+.. code-block:: sql
+
+    CREATE TABLE t (
+        id INTEGER,
+        start_ts TIMESTAMP,
+        end_ts TIMESTAMP,
+        PERIOD FOR my_period (start_ts, end_ts)
+    )
+
+Once created, a ``Period`` object is accessible via the tables's
+``Table.periods`` collection. (e.g. ``t.periods.my_period``.)
+
+
+Using Application-Time Periods in Primary Keys
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+SQL:2011 allows for including periods in composite primary or unique keys using
+the ``WITHOUT OVERLAPS`` clause. This can be acheived in SQLAlchemy in the same
+way as usual: either with the ``primary_key`` argument to :class:`Column`, or
```

Comment:
**Federico Caselli** (CaselIT) wrote:

This can be acheived in SQLAlchemy in the same
way as usual: either with the ``primary_key`` argument to :class:`Column`, or

This can be acheived in SQLAlchemy either with the ``primary_key`` argument to :class:`Period`, or

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803